File: pkg/service/search/search_service.go
/*
 * @Description: 搜索服务 - 搜索架构实现
 * @Author: 安知鱼
 * @Date: 2025-01-27 10:00:00
 * @LastEditTime: 2026-04-09
 * @LastEditors: 安知鱼
 */
package search

import (
	"context"
	"fmt"
	"log"
	"strings"

	"github.com/anzhiyu-c/anheyu-app/pkg/domain/model"
	"github.com/anzhiyu-c/anheyu-app/pkg/service/setting"
)

// AppSearcher 全局搜索器实例（可由插件管理器在启动时替换）
var AppSearcher model.Searcher

// SearchService 搜索服务
// 始终读取全局 AppSearcher 以支持插件热更新，不缓存本地引用
type SearchService struct {
	providers []SearchProvider
}

// SearchProvider 提供文章索引之外的公开内容搜索结果。
type SearchProvider interface {
	Search(ctx context.Context, query string, limit int) ([]*model.SearchHit, int64, error)
}

// NewSearchService 创建搜索服务实例
func NewSearchService() *SearchService {
	return &SearchService{}
}

// RegisterProvider 注册额外搜索内容提供者。
func (s *SearchService) RegisterProvider(provider SearchProvider) {
	if provider == nil {
		return
	}
	s.providers = append(s.providers, provider)
}

// Search 执行搜索
func (s *SearchService) Search(ctx context.Context, query string, page int, size int) (*model.SearchResult, error) {
	searcher := AppSearcher
	if searcher == nil {
		return nil, fmt.Errorf("搜索引擎未初始化")
	}
	result, err := searcher.Search(ctx, query, page, size)
	if err != nil {
		return nil, err
	}
	if result == nil {
		result = &model.SearchResult{}
	}
	if result.Pagination == nil {
		result.Pagination = &model.SearchPagination{Page: page, Size: size}
	}

	normalizeSearchHits(result.Hits)

	if strings.TrimSpace(query) == "" || len(s.providers) == 0 {
		return result, nil
	}

	totalExtra := int64(0)
	for _, provider := range s.providers {
		hits, total, err := provider.Search(ctx, query, size)
		if err != nil {
			return nil, fmt.Errorf("扩展搜索失败: %w", err)
		}
		normalizeSearchHits(hits)
		result.Hits = append(result.Hits, hits...)
		totalExtra += total
	}

	result.Pagination.Total += totalExtra
	result.Pagination.TotalPages = calculateTotalPages(result.Pagination.Total, result.Pagination.Size)
	return result, nil
}

func normalizeSearchHits(hits []*model.SearchHit) {
	for _, hit := range hits {
		if hit == nil || hit.Type != "" {
			continue
		}
		if hit.IsDoc {
			hit.Type = model.SearchHitTypeDoc
			if hit.URL == "" && hit.ID != "" {
				hit.URL = "/doc/" + hit.ID
			}
			continue
		}
		hit.Type = model.SearchHitTypePost
		if hit.URL == "" {
			targetID := hit.Abbrlink
			if targetID == "" {
				targetID = hit.ID
			}
			if targetID != "" {
				hit.URL = "/posts/" + targetID
			}
		}
	}
}

func calculateTotalPages(total int64, size int) int {
	if total <= 0 || size <= 0 {
		return 0
	}
	return int((total + int64(size) - 1) / int64(size))
}

// IndexArticle 索引文章
func (s *SearchService) IndexArticle(ctx context.Context, article *model.Article) error {
	searcher := AppSearcher
	if searcher == nil {
		log.Println("[警告] 搜索引擎未初始化，跳过索引操作")
		return nil
	}
	return searcher.IndexArticle(ctx, article)
}

// DeleteArticle 删除文章索引
func (s *SearchService) DeleteArticle(ctx context.Context, articleID string) error {
	searcher := AppSearcher
	if searcher == nil {
		log.Println("[警告] 搜索引擎未初始化，跳过删除索引操作")
		return nil
	}
	return searcher.DeleteArticle(ctx, articleID)
}

// RebuildAllIndexes 重建所有文章的搜索索引
func (s *SearchService) RebuildAllIndexes(ctx context.Context) error {
	searcher := AppSearcher
	if searcher == nil {
		log.Println("[警告] 搜索引擎未初始化，跳过重建索引操作")
		return fmt.Errorf("搜索引擎未初始化")
	}

	log.Println("开始重建搜索索引...")
	if err := searcher.ClearAllDocuments(ctx); err != nil {
		return fmt.Errorf("清理现有索引失败: %w", err)
	}

	log.Println("搜索索引清理完成，等待文章数据重建...")
	return nil
}

// InitializeSearchEngine 初始化搜索引擎（内置引擎降级链）
// 如果插件已提供搜索引擎（AppSearcher 已被设置），此函数不会覆盖
// 优先级: 插件搜索引擎 > Redis > Simple（内存）
func InitializeSearchEngine(settingSvc setting.SettingService) error {
	if AppSearcher != nil {
		log.Println("✅ 搜索引擎已由外部设置（可能来自插件），跳过内置引擎初始化")
		return nil
	}

	// 尝试使用 Redis 搜索模式
	redisSearcher, err := NewRedisSearcher(settingSvc)
	if err != nil {
		return fmt.Errorf("Redis 搜索初始化失败: %w", err)
	}

	if redisSearcher != nil {
		AppSearcher = redisSearcher
		log.Println("✅ Redis 搜索模式已启用")
		return nil
	}

	// 降级到简单搜索模式
	simpleSearcher := NewSimpleSearcher(settingSvc)
	AppSearcher = simpleSearcher
	log.Println("✅ 简单搜索模式已启用（降级方案）")
	return nil
}

File: server/src/search/search.service.ts
import { Inject, Injectable, Logger, OnModuleInit } from '@nestjs/common';
import { DRIZZLE } from '../database/database.module';
import { SettingsService } from '../settings/settings.service';
import { articles } from '../database/schemas/article.schema';
import { articlePostCategories } from '../database/schemas/article-post-category-pivot.schema';
import { articlePostTags } from '../database/schemas/article-post-tag-pivot.schema';
import { postCategories } from '../database/schemas/post-category.schema';
import { postTags } from '../database/schemas/post-tag.schema';
import { users } from '../database/schemas/user.schema';
import { generatePublicID, EntityType } from '../common/utils/sqids.util';
import { toISODateString } from '../common/utils/time.util';
import { eq, isNull, and, inArray, sql, desc } from 'drizzle-orm';

/**
 * SearchService implements FTS5 full-text search for articles.
 *
 * Per D-145: Uses SQLite FTS5 with contentless mode and unicode61 tokenizer.
 * Per D-147: bm25 ranking with title(10.0)/content(1.0)/keywords(5.0) weights.
 * Per D-150: FTS5 index rebuilt on startup from all published articles.
 * Per D-151: Exposes indexArticle/deleteArticle for ArticleService hooks.
 * Per D-152: Snippet extraction strips HTML and truncates to 150 chars.
 */
@Injectable()
export class SearchService implements OnModuleInit {
  private readonly logger = new Logger(SearchService.name);

  constructor(
    @Inject(DRIZZLE) private readonly db: any,
    private readonly settingsService: SettingsService,
  ) {}

  async onModuleInit(): Promise<void> {
    try {
      await this.ensureFts5Table();
    } catch (error) {
      this.logger.warn('FTS5 table initialization skipped (may not be available in test environment)');
    }
  }

  /**
   * Create FTS5 virtual table and rebuild indexes.
   * Per D-145: contentless mode with unicode61 tokenizer.
   * Per D-146: unicode61 with tokens "0" for basic CJK support.
   */
  async ensureFts5Table(): Promise<void> {
    this.logger.log('Creating FTS5 virtual table articles_fts...');

    // Create FTS5 virtual table with contentless mode
    await this.db.run(
      sql`CREATE VIRTUAL TABLE IF NOT EXISTS articles_fts USING fts5(
        title,
        content,
        keywords,
        content='',
        tokenize='unicode61 tokens 0'
      )`,
    );

    // Drop existing index data before rebuild for clean state
    await this.db.run(sql`DELETE FROM articles_fts`);

    // Rebuild indexes from all published articles
    await this.rebuildAllIndexes();

    this.logger.log('FTS5 virtual table created and indexes rebuilt');
  }

  /**
   * Rebuild FTS5 indexes from all published articles.
   * Per D-150: Full rebuild on startup ensures index consistency.
   */
  async rebuildAllIndexes(): Promise<void> {
    // Fetch all published, non-deleted articles
    const publishedArticles = await this.db
      .select({
        id: articles.id,
        title: articles.title,
        contentHtml: articles.contentHtml,
        keywords: articles.keywords,
      })
      .from(articles)
      .where(
        and(
          eq(articles.status, 'PUBLISHED'),
          isNull(articles.deletedAt),
        ),
      );

    this.logger.log(`Rebuilding FTS5 index for ${publishedArticles.length} articles...`);

    // Insert each article into FTS5
    for (const article of publishedArticles) {
      const plainContent = this.stripHtml(article.contentHtml || '');
      await this.db.run(
        sql`INSERT INTO articles_fts(rowid, title, content, keywords) VALUES (${article.id}, ${article.title}, ${plainContent}, ${article.keywords || ''})`,
      );
    }

    this.logger.log(`FTS5 index rebuilt with ${publishedArticles.length} articles`);
  }

  /**
   * Index a single article into FTS5.
   * Per D-151: Called from ArticleService on Create/Update.
   */
  async indexArticle(article: {
    id: number;
    title: string;
    contentHtml: string;
    keywords: string | null;
  }): Promise<void> {
    const plainContent = this.stripHtml(article.contentHtml || '');
    await this.db.run(
      sql`INSERT INTO articles_fts(rowid, title, content, keywords) VALUES (${article.id}, ${article.title}, ${plainContent}, ${article.keywords || ''})`,
    );
  }

  /**
   * Delete an article from FTS5 by database ID (rowid).
   * Per D-151: Called from ArticleService on Delete/Update.
   */
  async deleteArticle(articleDbId: number): Promise<void> {
    await this.db.run(
      sql`DELETE FROM articles_fts WHERE rowid = ${articleDbId}`,
    );
  }

  /**
   * Search articles using FTS5 with bm25 weighted ranking.
   * Per D-147: bm25(articles_fts, 10.0, 1.0, 5.0) for title/content/keywords.
   * Per D-148: Returns SearchResult matching Go model.SearchResult format.
   */
  async search(
    query: string,
    page: number,
    size: number,
  ): Promise<{
    pagination: {
      total: number;
      page: number;
      size: number;
      totalPages: number;
    };
    hits: any[];
  }> {
    // Empty query returns empty results (matches Go SimpleSearcher)
    if (!query || !query.trim()) {
      return {
        pagination: { total: 0, page, size, totalPages: 0 },
        hits: [],
      };
    }

    const offset = (page - 1) * size;

    try {
      // Get total matching count
      const countResult = await this.db.get(
        sql`SELECT count(*) as total FROM articles_fts WHERE articles_fts MATCH ${query}`,
      );
      const total = countResult?.total ?? 0;

      if (total === 0) {
        return {
          pagination: { total: 0, page, size, totalPages: 0 },
          hits: [],
        };
      }

      // Search with bm25 ranking
      const ftsResults = await this.db.all(
        sql`SELECT rowid as id, bm25(articles_fts, 10.0, 1.0, 5.0) AS rank
            FROM articles_fts
            WHERE articles_fts MATCH ${query}
            ORDER BY rank
            LIMIT ${size} OFFSET ${offset}`,
      );

      if (!ftsResults || ftsResults.length === 0) {
        return {
          pagination: { total, page, size, totalPages: Math.ceil(total / size) },
          hits: [],
        };
      }

      // Get article DB IDs from FTS5 results
      const articleDbIds = ftsResults.map((r: any) => r.id);

      // Fetch full article data with relations using Drizzle
      const articleRows = await this.db
        .select({
          id: articles.id,
          title: articles.title,
          contentHtml: articles.contentHtml,
          coverUrl: articles.coverUrl,
          abbrlink: articles.abbrlink,
          viewCount: articles.viewCount,
          wordCount: articles.wordCount,
          readingTime: articles.readingTime,
          isDoc: articles.isDoc,
          docSeriesId: articles.docSeriesId,
          keywords: articles.keywords,
          ownerId: articles.ownerId,
          copyrightAuthor: articles.copyrightAuthor,
          createdAt: articles.createdAt,
        })
        .from(articles)
        .where(inArray(articles.id, articleDbIds));

      // Build a map for quick lookup
      const articleMap = new Map<number, any>();
      for (const row of articleRows) {
        articleMap.set(row.id, row);
      }

      // Fetch categories and tags for all matched articles
      const categoriesMap = new Map<number, string>();
      const tagsMap = new Map<number, string[]>();

      if (articleDbIds.length > 0) {
        // Fetch categories
        const catRows = await this.db
          .select({
            articleId: articlePostCategories.articleId,
            name: postCategories.name,
          })
          .from(articlePostCategories)
          .innerJoin(
            postCategories,
            eq(articlePostCategories.postCategoryId, postCategories.id),
          )
          .where(
            and(
              inArray(articlePostCategories.articleId, articleDbIds),
              isNull(postCategories.deletedAt),
            ),
          );

        for (const row of catRows) {
          if (!categoriesMap.has(row.articleId)) {
            categoriesMap.set(row.articleId, row.name);
          }
        }

        // Fetch tags
        const tagRows = await this.db
          .select({
            articleId: articlePostTags.articleId,
            name: postTags.name,
          })
          .from(articlePostTags)
          .innerJoin(
            postTags,
            eq(articlePostTags.postTagId, postTags.id),
          )
          .where(
            and(
              inArray(articlePostTags.articleId, articleDbIds),
              isNull(postTags.deletedAt),
            ),
          );

        for (const row of tagRows) {
          if (!tagsMap.has(row.articleId)) {
            tagsMap.set(row.articleId, []);
          }
          tagsMap.get(row.articleId)!.push(row.name);
        }
      }

      // Build SearchHit objects in FTS5 rank order
      const hits = ftsResults.map((ftsRow: any) => {
        const article = articleMap.get(ftsRow.id);
        if (!article) return null;

        // Author: copyrightAuthor first, then site owner name from settings
        const author = article.copyrightAuthor ||
          this.settingsService.get('FRONT_DESK_SITE_OWNER_NAME') || '';

        // Doc series ID: encode via Sqids if non-null
        let docSeriesId = '';
        if (article.docSeriesId) {
          try {
            docSeriesId = generatePublicID(article.docSeriesId, EntityType.DocSeries);
          } catch {
            docSeriesId = '';
          }
        }

        return {
          id: generatePublicID(article.id, EntityType.Article),
          type: '',
          url: '',
          title: article.title,
          snippet: this.extractSnippet(article.contentHtml || ''),
          author,
          category: categoriesMap.get(article.id) || '',
          tags: tagsMap.get(article.id) || [],
          publish_date: toISODateString(article.createdAt),
          cover_url: article.coverUrl || '',
          abbrlink: article.abbrlink || '',
          view_count: article.viewCount ?? 0,
          word_count: article.wordCount ?? 0,
          reading_time: article.readingTime ?? 0,
          is_doc: article.isDoc ?? false,
          doc_series_id: docSeriesId,
        };
      }).filter((h: any) => h !== null);

      // Normalize hits: fill type and url per Go search_service.go
      const normalizedHits = this.normalizeSearchHits(hits);

      const totalPages = Math.ceil(total / size);

      return {
        pagination: { total, page, size, totalPages },
        hits: normalizedHits,
      };
    } catch (error) {
      // FTS5 table may not exist (e.g., in test environments or before migration)
      this.logger.warn('Search failed, FTS5 table may not exist');
      return {
        pagination: { total: 0, page, size, totalPages: 0 },
        hits: [],
      };
    }
  }

  /**
   * Extract snippet from HTML content.
   * Per D-152: Strip HTML tags, truncate to 150 chars with ellipsis.
   * Matches Go SimpleSearcher.articleToSearchHit snippet logic.
   */
  extractSnippet(contentHtml: string, maxLength = 150): string {
    // Strip HTML tags (same as Go's reHTMLTags.ReplaceAllString)
    const plainText = contentHtml.replace(/<[^>]*>/g, ' ').trim();
    // Collapse multiple spaces
    const collapsed = plainText.replace(/\s+/g, ' ');

    // Proper Unicode handling using spread for rune-level slicing
    const runes = [...collapsed];
    if (runes.length > maxLength) {
      return runes.slice(0, maxLength).join('') + '...';
    }
    return collapsed;
  }

  /**
   * Normalize search hits: fill type and url fields.
   * Per Go search_service.go lines 86-109:
   * - If type already set, skip
   * - If isDoc=true, type="doc" and url="/doc/{id}"
   * - Otherwise type="post" and url="/posts/{abbrlink}" (fallback to id)
   */
  normalizeSearchHits(hits: any[]): any[] {
    for (const hit of hits) {
      if (!hit || hit.type) {
        continue;
      }

      if (hit.is_doc) {
        hit.type = 'doc';
        if (!hit.url && hit.id) {
          hit.url = '/doc/' + hit.id;
        }
        continue;
      }

      hit.type = 'post';
      if (!hit.url) {
        const targetId = hit.abbrlink || hit.id;
        if (targetId) {
          hit.url = '/posts/' + targetId;
        }
      }
    }
    return hits;
  }

  /**
   * Strip HTML tags from content for FTS5 indexing.
   * Prevents HTML tag names from polluting search results.
   */
  private stripHtml(html: string): string {
    return html.replace(/<[^>]*>/g, ' ').replace(/\s+/g, ' ').trim();
  }
}
